[pruner] add cuda tests for pruner

kazhou opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #61993

Repeating `test_pruner` unit tests for Linear and Conv2d models with device = 'cuda' to confirm pruner will work on GPU
- set device to cuda
- move model to device
- assert that module.weight.device is cuda

Differential Revision: [D29829293](https://our.internmc.facebook.com/intern/diff/D29829293/)

kazhou committed:
[pruner] add cuda tests for pruner

Repeating `test_pruner` unit tests for Linear and Conv2d models with device = 'cuda' to confirm pruner will work on GPU
- set device to cuda
- move model to device
- assert that module.weight.device is cuda

Differential Revision: [D29829293](https://our.internmc.facebook.com/intern/diff/D29829293/)

[ghstack-poisoned]

supriyar commented on test/ao/sparsity/test_pruner.py:
```diff
@@ -150,6 +171,25 @@ def test_prepare_linear(self):
             assert type(module.parametrizations.weight[0]) == PruningParametrization
         assert model(x).shape == (128, 16)
 
+    def test_prepare_linear_cuda(self):
```
Do we need separate tests for all the `cuda` cases or can we parametrize the existing one to work on both devices?

z-a-f replied:
I agree -- we can wrap the test in a separate function:

```
def _test_prepare_linear_on_device(self, device):
    print(f'Testing on {device.type}')
    # All of your testing code here

def test_prepare_linear(self):
    self._test_prepare_linear_on_device(torch.device('cpu'))
    if torch.cuda.is_available():
        self._test_prepare_linear_on_device(torch.device('cuda'))
```

supriyar reviewed (changes requested):
I think we can try to avoid the code duplication here

z-a-f reviewed (changes requested):
some comments

kazhou committed:
Update on "[pruner] add cuda tests for pruner"


Repeating `test_pruner` unit tests for Linear and Conv2d models with device = 'cuda' to confirm pruner will work on GPU
- set device to cuda
- move model to device
- assert that module.weight.device is cuda

Differential Revision: [D29829293](https://our.internmc.facebook.com/intern/diff/D29829293/)

[ghstack-poisoned]

supriyar commented on test/ao/sparsity/test_pruner.py:
```diff
@@ -155,24 +140,14 @@ def test_constructor_cuda(self):
-            # Assume that this is the 1st/only parametrization
-            assert type(module.parametrizations.weight[0]) == PruningParametrization
-        assert model(x).shape == (128, 16)
+    def test_constructor(self):
+        self._test_constructor_on_device(torch.device('cpu'))
+        if torch.cuda.is_available():
+            self._test_constructor_on_device(torch.device('cuda'))
```
Can we use https://github.com/pytorch/pytorch/blob/master/test/quantization/core/test_workflow_module.py#L674 directly?

jerryzh168 replied:
yeah, also probably just loop through a list of devices and don't use hypothesis.

supriyar commented on test/ao/sparsity/test_pruner.py:
```diff
@@ -118,25 +118,10 @@ def update_mask(self, layer, **kwargs):
-        device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
-        # Cannot instantiate the base
+    def _test_constructor_on_device(self, device):
+        print(f'Testing on {device.type}')
```
Please remove print statements

jerryzh168 commented on test/ao/sparsity/test_pruner.py:
```diff
@@ -182,136 +213,201 @@ def test_prepare_linear_bias(self):
+        if torch.cuda.is_available():
+            self._test_convert_linear_on_device(torch.device('cuda'))
+
+    def _test_convert_linear_bias_on_device(self, device):
```
it looks like there is still a lot of duplications in these code, is it possible to simplify further?

kazhou committed:
Update on "[pruner] add cuda tests for pruner"


Repeating `test_pruner` unit tests for Linear and Conv2d models with device = 'cuda' to confirm pruner will work on GPU
- set device to cuda
- move model to device
- assert that module.weight.device is cuda

Differential Revision: [D29829293](https://our.internmc.facebook.com/intern/diff/D29829293/)

[ghstack-poisoned]

kazhou committed:
Update on "[pruner] add cuda tests for pruner"


Repeating `test_pruner` unit tests for Linear and Conv2d models with device = 'cuda' to confirm pruner will work on GPU
- set device to cuda
- move model to device
- assert that module.weight.device is cuda

Differential Revision: [D29829293](https://our.internmc.facebook.com/intern/diff/D29829293/)

[ghstack-poisoned]

jerryzh168 commented on test/ao/sparsity/test_pruner.py:
```diff
@@ -118,13 +121,49 @@ def update_mask(self, layer, **kwargs):
+            assert hasattr(module, "parametrizations")
+            # Assume that this is the 1st/only parametrization
+            assert type(module.parametrizations.weight[0]) == PruningParametrization
+        return True
```
nit: maybe rename this to something like _check_prepared_pruner` and remove the return, similar for other functions

jerryzh168 reviewed (approved):
Thanks for deduplicating the code!

kazhou committed:
Update on "[pruner] add cuda tests for pruner"


Repeating `test_pruner` unit tests for Linear and Conv2d models with device = 'cuda' to confirm pruner will work on GPU
- set device to cuda
- move model to device
- assert that module.weight.device is cuda

Differential Revision: [D29829293](https://our.internmc.facebook.com/intern/diff/D29829293/)

[ghstack-poisoned]

kazhou committed:
Update on "[pruner] add cuda tests for pruner"


Repeating `test_pruner` unit tests for Linear and Conv2d models with device = 'cuda' to confirm pruner will work on GPU
- set device to cuda
- move model to device
- assert that module.weight.device is cuda

Differential Revision: [D29829293](https://our.internmc.facebook.com/intern/diff/D29829293/)

[ghstack-poisoned]
